### ISSUE TITLE
"Could not find Chrome" error when running in IIS

### ISSUE THREAD

**Issue by anpool** [highcharts/node-export-server#446]
We have a node service API that uses highcharts-export-server (3.0.1) as a dependency. I have ran npm install and everything works fine when the node service is ran within Powershell, however if the service is ran in IIS (iisnode) then the following error is shown in the log when trying to make a request:

Mon Nov 20 2023 16:38:33 GMT+0100 [notice] - [browser] attempting to get a browser instance (try 10)
Mon Nov 20 2023 16:38:33 GMT+0100 [notice] - [browser] failed: Error: Could not find Chrome (ver. 119.0.6045.105). This can occur if either
 1. you did not perform an installation before running the script (e.g. `npm install`) or
 2. your cache path is incorrectly configured (which is: C:\WINDOWS\system32\config\systemprofile\.cache\puppeteer).
For (2), check out our guide on configuring puppeteer at https://pptr.dev/guides/configuration.

We have recently upgraded from highcharts-export-server version 2.1.0 which worked fine in IIS.

**Comment by jakubSzuminski** [comment]
Thanks for reporting @anpool!

Could you please try out the 3.0.3 version of the Export Server which has just been released and let me know if that issue still persist?

**Comment by anpool** [comment]
Thanks for the quick response. I tried upgrading to 3.0.3 but now the service no longer starts at all in Powershell:

Mon Nov 20 2023 17:00:56 GMT+0100 [notice] - [cache] Fetching and caching Highcharts dependencies.
Mon Nov 20 2023 17:00:56 GMT+0100 [notice] - [cache] Updating cache to Highcharts  
2023-11-20T16:00:56.264Z [info]: Highcharts service listening on port 5017
2023-11-20T16:00:56.294Z [error]: uncaughtException: Cannot read properties of undefined (reading 'map')
TypeError: Cannot read properties of undefined (reading 'map')
    at M (C:\projects\Spi\SpiLegacy\nodeservices\highcharts-service\node_modules\highcharts-export-server\dist\index.cjs:1:16280)
    at D (C:\projects\Spi\SpiLegacy\nodeservices\highcharts-service\node_modules\highcharts-export-server\dist\index.cjs:1:17795)
    at Object.initPool (C:\projects\Spi\SpiLegacy\nodeservices\highcharts-service\node_modules\highcharts-export-server\dist\index.cjs:1:41228)
    at Object.<anonymous> (C:\projects\Spi\SpiLegacy\nodeservices\highcharts-service\server.js:16:10)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47
    
    And when running in IIS the log shows the following:
    
    Mon Nov 20 2023 16:59:03 GMT+0100 [verbose] - [chart] Starting exporting process.
Mon Nov 20 2023 16:59:03 GMT+0100 [verbose] - [chart] Attempting to export from a raw input.
Mon Nov 20 2023 16:59:03 GMT+0100 [verbose] - [pool] Work received, starting to process.
Mon Nov 20 2023 16:59:03 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:03 GMT+0100 [verbose] - [pool] Acquiring worker
Mon Nov 20 2023 16:59:03 GMT+0100 [error] - [pool] Error when creating worker of an event id 34: Error creating page
Mon Nov 20 2023 16:59:03 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:03 GMT+0100 [error] - [pool] Error when creating worker of an event id 35: Error creating page
Mon Nov 20 2023 16:59:04 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:04 GMT+0100 [error] - [pool] Error when creating worker of an event id 36: Error creating page
Mon Nov 20 2023 16:59:04 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:04 GMT+0100 [error] - [pool] Error when creating worker of an event id 37: Error creating page
Mon Nov 20 2023 16:59:04 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:04 GMT+0100 [error] - [pool] Error when creating worker of an event id 38: Error creating page
Mon Nov 20 2023 16:59:04 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:04 GMT+0100 [error] - [pool] Error when creating worker of an event id 39: Error creating page
Mon Nov 20 2023 16:59:04 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:04 GMT+0100 [error] - [pool] Error when creating worker of an event id 40: Error creating page
Mon Nov 20 2023 16:59:05 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:05 GMT+0100 [error] - [pool] Error when creating worker of an event id 41: Error creating page
Mon Nov 20 2023 16:59:05 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:05 GMT+0100 [error] - [pool] Error when creating worker of an event id 42: Error creating page
Mon Nov 20 2023 16:59:05 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:05 GMT+0100 [error] - [pool] Error when creating worker of an event id 43: Error creating page
Mon Nov 20 2023 16:59:05 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:05 GMT+0100 [error] - [pool] Error when creating worker of an event id 44: Error creating page
Mon Nov 20 2023 16:59:06 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:06 GMT+0100 [error] - [pool] Error when creating worker of an event id 45: Error creating page
Mon Nov 20 2023 16:59:06 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:06 GMT+0100 [error] - [pool] Error when creating worker of an event id 46: Error creating page
Mon Nov 20 2023 16:59:06 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:06 GMT+0100 [error] - [pool] Error when creating worker of an event id 47: Error creating page
Mon Nov 20 2023 16:59:06 GMT+0100 [error] - [pool] Error creating a new page in pool entry creation! invalid page
Mon Nov 20 2023 16:59:06 GMT+0100 [error] - [pool] Error when creating worker of an event id 48: Error creating page
Mon Nov 20 2023 16:59:06 GMT+0100 [error] - [pool] Error when acquiring worker of an event id 33: Error

**Comment by jakubSzuminski** [comment]
Thanks for sharing!

1. Would it be possible for you to test the same config but by using **ESModule** (`import` statement)? Currently, I see that you're using `require` to get the Export Server. This would help us identify whether the problem is related to the entire Export Server or just the `.cjs` build.

2. Could you please share any helpful information about your: environment, setup, command you're running that may help us in debugging this issue?

Thanks in advance :)

**Comment by anpool** [comment]
I've converted the service to ESModule syntax but still the same issues, 1.0.1 only works in Powershell and not IIS, and 1.0.3 works in neither.

**Environment:**
OS: Windows 10 Enterprise
Node: v18.12.1
IIS: 10
iisnode

Below are some samples of the code from the server.js and an endpoint using the highcharts-export-server:

**Server.mjs:**

    import fs from 'fs';
    import path from 'path';
    import bodyParser from 'body-parser';
    import express from 'express';
    import morgan from 'morgan';
    import logger from './config/winston.mjs';
    import exporter from 'highcharts-export-server';
    
    const logDir = 'log';
    if (!fs.existsSync(logDir)) {
        // Create the directory if it does not exist
        fs.mkdirSync(logDir);
    }
    
    const app = express();
    
    // Set up a pool of PhantomJS workers
    exporter.initPool();
    
    const accessLogStream = fs.createWriteStream(
        path.join(path.resolve(), logDir, 'access.log'),
        { flags: 'a' }
    );
    
    morgan.token('date', () => new Date().toISOString());
    
    const morganFormat = (tokens, req, res) => [
        tokens.date(req, res),
        tokens.method(req, res),
        tokens.url(req, res),
        tokens.status(req, res),
        tokens.res(req, res, 'content-length'), '-',
        tokens['response-time'](req, res), 'ms'
    ].join(' ');
    
    app.use(morgan(morganFormat, { stream: accessLogStream }));
    app.use(bodyParser.json()); // Support json encoded bodies
    app.use(bodyParser.urlencoded({ extended: true })); // Support encoded bodies
    
    import { version, monitorUnit } from './version.mjs';
    

    // ... rest of the code for routing api
    

    app.set('env', 'development');
    app.disable('x-powered-by');
   

     // ... rest of the code for listening on a port
    

    // Gracefully shut down the pool when the application exits
    process.on('SIGINT', () => {
        exporter.killPool();
        process.exit();
    });
    
    export default app;


**Endpoint:**

      var express = require('express')
      var router = express.Router()
      import exporter from 'highcharts-export-server';
      
      // ... rest of the code
        
      router.post('/', async function (req, res) {
  
        // ... rest of the code
    
        try {
    
            // ... rest of the code
    
            //Export settings
            const exportSettings = {
                export: {
                    fileName: 'Chart Name',     // The name of the chart
                    file: 'chart-name.' + type, // The name of the chart plus its extension
                    type: type,
                    options: options
                }
            };
    
            // Perform an export
              exporter.startExport(exportSettings, function (response, err) {
                  if (err) {
                      return commonUtils.handleErrorResponse(res, err);
                  } else {
                      commonUtils.handleExportResponse(res, response, type)
                  }
              })
          } catch (error) {
              return commonUtils.handleErrorResponse(res, error)
          }
      })

**Comment by jakubSzuminski** [comment]
One thing that comes to my mind is the improper use of the `initPool` method. This method has changed in our Puppeteer-based version of the Export Server and it now requires `await` as you can see in this part of our readme: https://github.com/highcharts/node-export-server/tree/stable#using-as-a-nodejs-module

```
const exportSettings = {
  export: {
    type: 'png',
    options: {
      series: [
        {
          type: 'line',
          data: [1, 3, 2, 4]
        }
      ]
    }
  }
};

// Set the new options and merge it with the default options
const options = exporter.setOptions(exportSettings);

// Initialize a pool of workers
await exporter.initPool(options);

// Perform an export
exporter.startExport(exportSettings, function (res, err) {
  // The export result is now in res.
  // It will be base64 encoded (res.data).

  // Kill the pool when we're done with it.
  exporter.killPool();
});
```

Please try using `await exporter.initPool`. If it does not fix the issue, I will further investigate what's going on.

**Comment by anpool** [comment]
Changed to await but still no change unfortunately

**Comment by DesarrolloS-B** [comment]
+1

**Comment by anpool** [comment]
Any luck with this?

**Comment by anpool** [comment]
Thanks for the update, could it be related to the removal of the environmental variables specifying the npm install options (e.g Highcharts version, gantt etc)? It appeared that these were no longer needed for version 3 and were therefore removed.

**Comment by jakubSzuminski** [comment]
I don't think so, you can take a look at this file: https://github.com/highcharts/node-export-server/blob/stable/lib/schemas/config.js where there is a `defaultConfig` defined. If no envs exist, the values from this JSON config are taken into account. We have set all the necessary default values so it should not be a problem.

**Comment by jakubSzuminski** [comment]
@anpool could you please share some reproduction steps that we could follow to throw this error locally? How did you install the Export Server? Could you give us a minimal piece of code in `Node.js + Express` that would trigger this?

**Comment by anpool** [comment]
We're you able to reproduce the issue with the example project?

**Comment by jakubSzuminski** [comment]
Hi @anpool, not yet. As you've mentioned, after running `npm start` everything works as expected. Only after running it on `IIS` something is incorrect which indicates that a proper configuration of the **IIS** is needed. As I'm not an expert in IIS, I will have to dive deeper into this topic after all the prioritized bugs & features on this repository are resolved.

---
In the meantime, here are some tips that you may find helpful in case you want to try to solve this problem:

1. Puppeteer requires a specific version of Chromium to be available, and it seems like it's either not installed or not found in the expected location when the server is running under **IIS**.
2. Puppeteer typically downloads its own version of Chromium when it's installed. Verify that this has happened and that Chromium is available in the expected path. The default path is usually within the `node_modules/puppeteer/.local-chromium` directory of your project.
3. The error message indicates an issue with the cache path `(C:\WINDOWS\system32\config\systemprofile\.cache\puppeteer)`. This path might not be accessible or suitable when running under IIS. Consider configuring Puppeteer to use a different cache path. You can set the `PUPPETEER_CACHE_DIR` environment variable to a more suitable location.
4. When running under IIS, the application might have different permissions or environment settings compared to when it's run directly. Ensure that the IIS application pool has the necessary permissions to access the directories where Puppeteer and Chromium are located. Also, check if the environment variables are correctly set for the IIS process.
5. If the automatic download of Chromium doesn't work well with your IIS setup, you can manually specify the path to a Chromium or Chrome executable that Puppeteer should use. This can be done by setting the `executablePath` option in Puppeteer's launch configuration.

**Comment by anpool** [comment]
Ok thanks for your suggestions ill take a look at these. I had been suspecting that it maybe a permissions issue. Just to confirm, you were not able to reproduce on IIS, it worked for you?

**Comment by anpool** [comment]
Managed to fix the issue, it was a permissions issue with the cache path as you had suggested. Fixed it by setting a different path to a more suitable location and giving IIS the necessary permissions to the directory. Thanks for your help

**Comment by jakubSzuminski** [comment]
Great, I'm glad to hear that!
